### PR TITLE
Fix multiple print of `precompiling project...`

### PR DESCRIPTION
**Pull request by ianshmean**
Not seen on my system, but by others. It seems `printpkgstyle` has a yield point or two, so better to set the bool before

### CODE REVIEW

**Comment by KristofferC** [comment]
>  It seems printpkgstyle has a yield point or two

It prints so that sounds very likely.

**Comment by giordano** [comment]
I confirm that with this PR I see only one "Precompiling project..." message, although still a bit messy:
```
julia> Pkg.precompile()
Precompiling[ Info: Precompiling IteratorInterfaceExtensions [82899510-4779-5014-852e-03e436cf321d]
[ Info: Precompiling Requires [ae029012-a4dd-5104-9daa-d747884805df]
[ Info: Precompiling hicolor_icon_theme_jll [059c91fe-1bad-52ad-bddd-f7b78713c282]
[ Info: Precompiling Parsers [69de0a69-1ddd-5017-9359-2bf0b02dc9f0]
[ Info: Precompiling Adapt [79e6a3ab-5dfb-504d-930d-738a2a938a0e]
[ Info: Precompiling JLLWrappers [692b3bcd-3c85-4b1f-b108-f13ce0eb3210]
[ Info: Precompiling StaticArrays [90137ffa-7385-5640-81b9-e52037218182]
[ Info: Precompiling Grisu [42e2da0e-8278-4e71-bc24-59509adca0fe]
 project...
```

File: src/API.jl [hunk 1]
```diff
@@ -975,7 +975,10 @@ function precompile(ctx::Context; internal_call::Bool=false)
                 end
                 is_direct_dep =  pkg in direct_deps
                 try
-                    !any(values(was_recompiled)) && printpkgstyle(ctx, :Precompiling, "project...")
+                    if !any(values(was_recompiled))
+                        was_recompiled[pkg] = true # needed to prevent `@async` race, and multiple prints
+                        printpkgstyle(ctx, :Precompiling, "project...")
```
Comment (KristofferC) [inline src/API.jl:981]: The tests in Base use a lock for printing:

https://github.com/JuliaLang/julia/blob/f6d1032763cfa160c117f0a5de25ba3370a3dca4/test/runtests.jl#L110

**Comment by giordano** [comment]
```
julia> Pkg.precompile()
ERROR: UndefVarError: Event not defined
Stacktrace:
 [1] precompile(ctx::Pkg.Types.Context; internal_call::Bool)
   @ Pkg.API /tmp/Pkg.jl/src/API.jl:949
 [2] #precompile#195
   @ /tmp/Pkg.jl/src/API.jl:913 [inlined]
 [3] precompile()
   @ Pkg.API /tmp/Pkg.jl/src/API.jl:913
 [4] top-level scope
   @ REPL[3]:1
```

**Comment by giordano** [comment]
```
julia> Pkg.precompile()
Precompiling project...
[ Info: Precompiling MbedTLS_jll [c8ffd9c3-330d-5841-b78e-0817d7145fa1]
Precompiling project...
[ Info: Precompiling IteratorInterfaceExtensions [82899510-4779-5014-852e-03e436cf321d]
Precompiling project...
[ Info: Precompiling Requires [ae029012-a4dd-5104-9daa-d747884805df]
Precompiling project...
[ Info: Precompiling hicolor_icon_theme_jll [059c91fe-1bad-52ad-bddd-f7b78713c282]
Precompiling project...
[ Info: Precompiling Parsers [69de0a69-1ddd-5017-9359-2bf0b02dc9f0]
Precompiling project...
[ Info: Precompiling Adapt [79e6a3ab-5dfb-504d-930d-738a2a938a0e]
Precompiling project...
[ Info: Precompiling JLLWrappers [692b3bcd-3c85-4b1f-b108-f13ce0eb3210]
Precompiling project...
[ Info: Precompiling StaticArrays [90137ffa-7385-5640-81b9-e52037218182]
Precompiling project...
```
almost :stuck_out_tongue:

**Comment by ianshmean** [comment]
Where's the yield?!
```
if !any(values(was_recompiled))
                        lock(print_lock) do
                            printpkgstyle(ctx, :Precompiling, "project...")
                        end
                    end
                    was_recompiled[pkg] = true
```

**Comment by giordano** [comment]
The [latest version](https://github.com/JuliaLang/Pkg.jl/pull/2047/commits/7d5e45183874184df44e900e83e9fc3066e9ade8) works well!
```
julia> Pkg.precompile()
Precompiling project...
[ Info: Precompiling MbedTLS_jll [c8ffd9c3-330d-5841-b78e-0817d7145fa1]
[ Info: Precompiling IteratorInterfaceExtensions [82899510-4779-5014-852e-03e436cf321d]
[ Info: Precompiling Requires [ae029012-a4dd-5104-9daa-d747884805df]
[ Info: Precompiling hicolor_icon_theme_jll [059c91fe-1bad-52ad-bddd-f7b78713c282]
[ Info: Precompiling Parsers [69de0a69-1ddd-5017-9359-2bf0b02dc9f0]
[ Info: Precompiling Adapt [79e6a3ab-5dfb-504d-930d-738a2a938a0e]
[ Info: Precompiling JLLWrappers [692b3bcd-3c85-4b1f-b108-f13ce0eb3210]
[ Info: Precompiling StaticArrays [90137ffa-7385-5640-81b9-e52037218182]
[ Info: Precompiling Grisu [42e2da0e-8278-4e71-bc24-59509adca0fe]
[...]
```

**Comment by ianshmean** [comment]
Great!

**Comment by ianshmean** [comment]
Merge? @KristofferC

**Comment by tkf** [comment]
The CI is broken because we are calling the method of `compilecache` that does not exist anymore. #2048 should fix it.

**Comment by ianshmean** [comment]
Great. Can we merge? I want to see if master fixes #2057

**Pull request merged**